### PR TITLE
Add device runtime API for the plug-in to register platform python module into torch

### PR DESCRIPTION
**Pull request by chengjunlu**
## Motivation
Allow the out-of-tree Pytorch plug-in, for the device type other than CUDA, to add the runtime interface to the `torch` module. The runtime interface of the device can be referred with the device type name in the `torch` module. I.E., `torch.cuda` or `torch.xpu`.

## Solution
- Add a register interface for the plug-in to add the platform python module into `torch` module with the device type name. I.E., The `torch.xpu` can be used to refer the XPU runtime interface after the XPU runtime module is registered with `torch._register_device_module('xpu', xpu_module)` in Intel's XPU plug-in.

## Additional Context
More details about runtime has been discussed in #53707.

### CODE REVIEW

**Comment by chengjunlu** [comment]
@ezyang @gujinghui 
Python API for register the device module into `torch`

**Comment by ezyang** [comment]
OK, I am happy now that this PR is short.

Some pieces that seem missing:

1. Don't you need a way of getting device modules given a device type? Seems like it would be good to set this up and make sure CUDA is registered into it correctly
2. Error checking to make sure we don't override already existing modules
3. Tests

**Comment by chengjunlu** [comment]
> OK, I am happy now that this PR is short.
> 
> Some pieces that seem missing:
> 
> 1. Don't you need a way of getting device modules given a device type? Seems like it would be good to set this up and make sure CUDA is registered into it correctly
I'd like to add those changes along with the DDP CUDA decouple changes. We can have a full view of the changes required. 

> 2. Error checking to make sure we don't override already existing modules
> 3. Tests
I will update this PR.

**Comment by gujinghui** [comment]
@ezyang  Could you help review it? Thanks.

File: test/test_utils.py [hunk 1]
```diff
@@ -825,5 +825,33 @@ def test_load_standalone(self):
             shutil.rmtree(build_dir)
 
 
+class DummyXPUModule(object):
+    @staticmethod
+    def is_avaliable():
```
Comment (ezyang) [inline test/test_utils.py:830]: available

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**